packages/integrations/node/src/server.ts

```diff
@@ -0,0 +1,36 @@
+import type { SSRManifest } from 'astro';
+import type { IncomingMessage, ServerResponse } from 'http';
+import { NodeApp } from 'astro/app/node';
+import { polyfill } from '@astrojs/webapi';
+
+polyfill(globalThis);
+
+export function createExports(manifest: SSRManifest) {
+	const app = new NodeApp(manifest, new URL(import.meta.url));
+	return {
+		async handler(req: IncomingMessage, res: ServerResponse, next?: (err?: unknown) => void) {
+			const route = app.match(req);
+
+			if(route) {
+				try {
+					/** @type {Response} */
+					const response = await app.render(req);
+					const html = await response.text();
+					res.writeHead(response.status, {
+						'Content-Type': 'text/html; charset=utf-8',
+						'Content-Length': Buffer.byteLength(html, 'utf-8'),
+					});
```

Comment:
Don't we also need to support non-HTML routes like `src/pages/endpoint.json.js`?

Reply:
Yep, fixed this.